yarn-project/prover-client/src/orchestrator/orchestrator.ts

```diff
@@ -0,0 +1,515 @@
+    } satisfies CircuitSimulationStats);
+    if (!this.provingState?.verifyState(stateIdentifier)) {
+      return;
+    }
+    logger.info(`Completed merge rollup at level ${level}, index ${index}`);
+    this.storeAndExecuteNextMergeLevel(level, index, circuitOutputs, stateIdentifier);
+  }
+
+  // Executes the root rollup circuit
+  private async runRootRollup(
+    mergeInputData: MergeRollupInputData,
+    rootParityInput: RootParityInput,
+    stateIdentifier: string,
+  ) {
+    const [circuitsOutput, proof] = await executeRootRollupCircuit(
+      [mergeInputData.inputs[0]!, mergeInputData.proofs[0]!],
+      [mergeInputData.inputs[1]!, mergeInputData.proofs[1]!],
+      rootParityInput,
+      this.provingState!.newL1ToL2Messages,
+      this.simulator,
+      this.prover,
+      this.db,
+      logger,
+    );
+    logger.info(`Completed root rollup`);
+    // Collect all new nullifiers, commitments, and contracts from all txs in this block
+    const txEffects: TxEffect[] = this.provingState!.allTxs.map(tx => toTxEffect(tx));
+
+    const blockBody = new Body(
+      padArrayEnd(this.provingState!.newL1ToL2Messages, Fr.ZERO, NUMBER_OF_L1_L2_MESSAGES_PER_ROLLUP),
+      txEffects,
+    );
+
+    const l2Block = L2Block.fromFields({
+      archive: circuitsOutput.archive,
+      header: circuitsOutput.header,
+      body: blockBody,
+    });
+
+    if (!l2Block.body.getTxsEffectsHash().equals(circuitsOutput.header.contentCommitment.txsEffectsHash)) {
+      logger(inspect(blockBody));
+      throw new Error(
+        `Txs effects hash mismatch, ${l2Block.body
+          .getTxsEffectsHash()
+          .toString('hex')} == ${circuitsOutput.header.contentCommitment.txsEffectsHash.toString('hex')} `,
+      );
+    }
+
+    const provingResult: ProvingResult = {
+      block: l2Block,
+      proof,
+    };
+    logger.info(`Successfully proven block ${l2Block.number}!`);
+    this.provingState!.resolve(provingResult, stateIdentifier);
+  }
+
+  // Executes the base parity circuit and stores the intermediate state for the root parity circuit
+  // Enqueues the root parity circuit if all inputs are available
+  private async runBaseParityCircuit(inputs: BaseParityInputs, index: number, stateIdentifier: string) {
+    const [duration, circuitOutputs] = await elapsed(() =>
+      executeBaseParityCircuit(inputs, this.simulator, this.prover, logger),
+    );
+    logger.debug(`Simulated base parity circuit`, {
+      eventName: 'circuit-simulation',
+      circuitName: 'base-parity',
+      duration,
+      inputSize: inputs.toBuffer().length,
+      outputSize: circuitOutputs.toBuffer().length,
+    } satisfies CircuitSimulationStats);
+    if (!this.provingState?.verifyState(stateIdentifier)) {
+      return;
+    }
+    this.provingState!.setRootParityInputs(circuitOutputs, index);
+
+    if (!this.provingState!.areRootParityInputsReady()) {
+      // not ready to run the root parity circuit yet
+      return;
+    }
```

Comment:
I'm having a hard time following the flow here: who takes care of retrying this once the root parity inputs are ready?